### PR TITLE
Normative: Change the hourCycle default logic

### PR DESCRIPTION
**Pull request by FrankYFTang**
While hour12 is  false, let hourCycle to be alwasyt 'h23' but not 'h24'.

The current logic is not reasonable. We see no region in the CLDR use  h24 hour cycle as default.  While we set a hour12: false on a 12 hour system region (which use 1:00 - 12:59), we should set to h23 instead of h24

We need to make this change
so while hour12 is false on 'en' locale, it will be set to h23 instead of h24
but also while hour12 is true on 'ja' locale, it will be set to h11 but not h12.

Address https://github.com/tc39/ecma402/issues/402

### CODE REVIEW

**Comment by FrankYFTang** [comment]
notice grep "<hours.*allowed" ./cldr/common/supplemental/supplementalData.xml show possibility of h H or K but no possibility of k in the latest CLDR and it is unlikely that will happen in the future from my understanding.

**Review by ryzokuken — APPROVED** [review]

File: spec/datetimeformat.html
Comment (sffc) [inline spec/datetimeformat.html:83]: Why are you changing it here but not 2 lines above?

Comment (sffc) [inline spec/datetimeformat.html:83]: Would better logic be,

```
1. If _hour12_ is *true*, then
  1. If _hcDefault_ is *"h23"* or *"h24"*, let _hc_ be *"h12"*. Otherwise, let _hc_ be _hcDefault_.
1. Else if _hour12_ is *false*, then
  1. If _hcDefault_ is *"h11"* or *"h12"*, let _hc_ be *"h23"*. Otherwise, let _hc_ be _hcDefault_.
```

**Review by sffc — CHANGES_REQUESTED** [review]

File: spec/datetimeformat.html
Comment (FrankYFTang) [reply spec/datetimeformat.html:83]: because if I change the 2 line above, then ja will not work
see https://github.com/tc39/ecma402/issues/402#issuecomment-623628320
see the following code
```js
(new Intl.DateTimeFormat("ja", {hour: "numeric", hour12: true})).resolvedOptions().hourCycle;
// h11
(new Intl.DateTimeFormat("en", {hour: "numeric", hour12: true})).resolvedOptions().hourCycle
// h12
(new Intl.DateTimeFormat("fr", {hour: "numeric", hour12: true})).resolvedOptions().hourCycle
// h12
```
and this should be kept.

Comment (FrankYFTang) [reply spec/datetimeformat.html:83]: this will not work for
```js
(new Intl.DateTimeFormat("ja", {hour: "numeric", hour12: true})).resolvedOptions().hourCycle;
```
_hcDefault_ for ja is h23 but if hour12:true it should be h11, not h23 nor h12
your logic will make it h23.
also there are more problem for the case if _hour12_ is *false*

Comment (sffc) [reply spec/datetimeformat.html:83]: Can you cite the CLDR data that shows that "ja" in 12-hour mode is expected to use h11, and likewise any locale in 24-hour mode that is expected to use h24?

Comment (FrankYFTang) [reply spec/datetimeformat.html:83]: > Can you cite the CLDR data that shows that "ja" in 12-hour mode is expected to use h11, and likewise any locale in 24-hour mode that is expected to use h24?

That is not what my PR proposed.

What my PR proposed is 
the following mapping

| hour12  | undefined | false  |   true |  
|---|---|---|---| 
| h11  |  h11  | h23 |  h11  | 
| h12 | h12  |  h23* |  h12  |
| h23  | h23  |  h23 |  h11 |
| h24  |  h24  |  h24 |  h12  |

The differences from the status quo is while the hour cycle is h12 and hour12:false  instead of making it h24 (which show 24:30:59) we show (00:30:59) and that impact mostly "en-US" or "en-CA" , "fr-CA" locales

Comment (FrankYFTang) [reply spec/datetimeformat.html:83]: about japanese, I am not sure this count as what you asked
https://github.com/unicode-org/cldr/blob/main/common/main/ja.xml#L3059
```xml
						<dateFormatItem id="h">aK時</dateFormatItem>
						<dateFormatItem id="H">H時</dateFormatItem>
```
h-> aK =>  use h11 when asking for not in 24 hour mode
H->H => use h23 when asking for in 24 hour mode

Comment (sffc) [reply spec/datetimeformat.html:83]: Thanks for the Japanese citation

I'm not super happy with the spec codifying this table. The only restriction should be that you get either h11 or h12 in 12-hour mode or either h23 or h24 in 24-hour mode. Which one should be ILD.

Comment (FrankYFTang) [reply spec/datetimeformat.html:83]: I understand why you are not happen with the spec codifying this table. 

" The only restriction should be that you get either h11 or h12 in 12-hour mode or either h23 or h24 in 24-hour mode"
YES, but that is not what the current spec text specify.  The current spec text specifiy a logic  to resolve that instead purely depend on the  CLDR to resolve that.

Comment (anba) [reply spec/datetimeformat.html:83]: Alternatively we could add two new fields to [LocaleData](https://tc39.es/ecma402/#sec-intl.datetimeformat-internal-slots) to describe the preferred hour-cycles for 12- and 24-hour clocks.

> - [[LocaleData]].[[\<locale\>]].[[hourCycle12]] must be a String value equal to "h11" or "h12". 
> - [[LocaleData]].[[\<locale\>]].[[hourCycle24]] must be a String value equal to "h23" or "h24". 


And then rewrite steps 24-26 as:
> 25. If hour12 is true, then
>     a. Let hc be dataLocaleData.[[hourCycle12]].
> 26. Else if hour12 is false, then
>     a. Let hc be dataLocaleData.[[hourCycle24]].
> 27. Else,
>     a. Assert: hour12 is undefined.
>     b. Let hc be r.[[hc]].
>     c. If hc is null, set hc to dataLocaleData.[[hourCycle]].

Comment (FrankYFTang) [reply spec/datetimeformat.html:83]: > Alternatively we could add two new fields to [LocaleData](https://tc39.es/ecma402/#sec-intl.datetimeformat-internal-slots) to describe the preferred hour-cycles for 12- and 24-hour clocks.
> 
> >

hum.... that will make it more a data driven approach ( which is good). Shane- what do you think?

Comment (sffc) [reply spec/datetimeformat.html:83]: +1 on @anba's approach.

**Comment by FrankYFTang** [comment]
PTAL I took anba's suggestion and make it data driven.

**Review by sffc — APPROVED** [review]

**Review by sffc — APPROVED** [review]

**Review by ryzokuken — APPROVED** [review]

**Comment by sffc** [comment]
TG2 agreement: https://github.com/tc39/ecma402/blob/master/meetings/notes-2023-04-06.md#normative-change-the-hourcycle-default-logic-758

**Comment by FrankYFTang** [comment]
somehow no one ever present this PR to TG39 in May 2023 or July 2023 meeting. I add it to https://github.com/tc39/agendas/blob/main/2023/09.md

**Comment by FrankYFTang** [comment]
we need test262 tests for this PR

**Comment by jufemaiz** [comment]
I'm happy to help out with this if needed @FrankYFTang. What's the requirement here?

**Comment by FrankYFTang** [comment]
@anba - any suggetion how should we write test262 test to test this PR?

**Comment by anba** [comment]
> @anba - any suggetion how should we write test262 test to test this PR?

<https://github.com/tc39/test262/blob/main/test/intl402/DateTimeFormat/prototype/resolvedOptions/hourCycle-default.js> needs to be updated to reflect the PR changes. The test is already marked as locale-specific, so we could simply change it to:
```js
let locales = ["en", "fr", "it", "ja", "zh", "ko", "ar", "hi"];

for (let locale of locales) {
  let hcDefault = new Intl.DateTimeFormat(locale, {hour: "numeric"}).resolvedOptions().hourCycle;
  assert(
    hcDefault === "h11" || hcDefault === "h12" || hcDefault === "h23" || hcDefault === "h24",
    "hcDefault is one of [h11, h12, h23, h24]"
  );

  let hour12 = new Intl.DateTimeFormat(locale, {hour: "numeric", hour12: true}).resolvedOptions().hourCycle;
  assert(hour12 === "h11" || hour12 === "h12", "hour12 is one of [h11, h12]");

  let hour24 = new Intl.DateTimeFormat(locale, {hour: "numeric", hour12: false}).resolvedOptions().hourCycle;
  assert(hour24 === "h23" || hour24 === "h24", "hour24 is one of [h23, h24]");

  if (hcDefault === "h11" || hcDefault === "h12") {
    assert.sameValue(hour12, hcDefault, "hour12 matches hcDefault");
  } else {
    assert.sameValue(hour24, hcDefault, "hour24 matches hcDefault");
  }

  // 24-hour clock uses the "h23" format in all locales.
  assert.sameValue(hour24, "h23");

  // 12-hour clock uses the "h12" format in all locales except "ja".
  assert.sameValue(hour12, locale === "ja" ? "h11" : "h12");
}
```

**Comment by ryzokuken** [comment]
@FrankYFTang @anba is merging this PR blocked on the tests?

**Comment by ryzokuken** [comment]
2023-09-26: This PR achieved TG1 consensus

**Comment by FrankYFTang** [comment]
This PR reach concensus on 2023-09-26 meeting

**Comment by FrankYFTang** [comment]
I believe the PR need to merge first before merging the test, right?

**Comment by FrankYFTang** [comment]
> @FrankYFTang @anba is merging this PR blocked on the tests?

The test is blocked by the merge of this PR https://github.com/tc39/test262/pull/3885

File: spec/datetimeformat.html
Comment (anba) [inline spec/datetimeformat.html:208]:
```suggestion
          [[LocaleData]].[[&lt;_locale_&gt;]].[[hourCycle12]] must be a String value equal to *"h11"* or *"h12"*.
```

File: spec/datetimeformat.html
Comment (anba) [inline spec/datetimeformat.html:211]:
```suggestion
          [[LocaleData]].[[&lt;_locale_&gt;]].[[hourCycle24]] must be a String value equal to *"h23"* or *"h24"*.
```

**Review by anba — APPROVED** [review]
Two small issues in the description text, otherwise LGTM!

**Comment by jufemaiz** [comment]
🎉

**Comment by jufemaiz** [comment]
Can I assume that this will form part of a changeset for the 10th edition?

Ref: https://ecma-international.org/publications-and-standards/standards/ecma-402/

**Comment by jufemaiz** [comment]
> https://ecma-international.org/publications-and-standards/standards/ecma-402/

FTR I believe it first was seen in v11.0 (2024-06)

**Pull request merged**
